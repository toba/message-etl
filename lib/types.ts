export enum Source {
   FacebookMessenger,
   GoogleVoice,
   GMail
}

export enum Relation {
   None = 0,
   Self,
   Other
}

export interface Message {
   /** Who sent the message. */
   from: Relation;
   /** When the message was sent. */
   on: Date;
   subject?: string;
   /** The platform used to convey the message. */
   source: Source;
   /** Content of the message. */
   text: string | null;
   /** An emoji reaction. */
   reaction?: string;
}

/**
 * Methods to process and parse messages from a particular platform.
 */
export interface Reader {
   sourceName: string;
   /** Filter file list so it only includes those for the platform. */
   filter: (fileName: string) => boolean;
   /** Process an exported file. */
   process: (fileText: string) => Message[];
   /** Parse a single message from the exported file. */
   parse: (msg: any) => Message | null;
}

export interface Writer {
   serialize: (m: Message) => string;
   save: (messages: Message[]) => boolean;
}

export interface Person {
   name: string;
   alias: string[];
   email: string[];
   phone: string[];
}

export type People = { [key: number]: Person };

export namespace Facebook {
   export interface Reaction {
      /** Emoji */
      reaction: string;
      /** Person's name */
      actor: string;
   }

   export interface Link {
      uri: string;
   }

   export interface MediaLink extends Link {
      creation_timestamp: number;
   }

   export interface VideoLink extends MediaLink {
      thumbnail: Link;
   }

   export interface Message {
      sender_name: string;
      timestamp_ms: number;
      content: string;
      type: 'Generic';
      sticker?: Link;
      videos?: VideoLink[];
      photos?: MediaLink[];
      reactions?: Reaction[];
   }

   interface Participant {
      name: string;
   }

   /**
    * Exported Facebook Messenger data. Export files are per-conversation.
    * @see https://www.facebook.com/help/1701730696756992?helpref=hc_global_nav
    */
   export interface Conversation {
      participants: Participant[];
      messages: Message[];
      title: string;
      is_still_participant: boolean;
      thread_type: string;
      thread_path: string;
   }
}
