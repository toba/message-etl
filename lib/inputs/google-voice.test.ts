import '@toba/test';
import { readFileText } from '@toba/test';
import { parse as parseHTML } from 'node-html-parser';
import { gv } from './index';
import { Source, Relation } from '../types';

jest.mock('../config/index');

const name = 'Other Hyphen-Name';
const raw = `<div class="message">
   <abbr class="dt" title="2013-06-17T20:46:00.861-04:00">Jun 17, 2013, 8:46:00 PM Eastern Time</abbr>:
      <cite class="sender vcard">
         <a class="tel" href="tel:+12087242867">
            <span class="fn">First Last-Hyphen</span>
         </a>
      </cite>:
      <q>Flight 2244, 15 minutes delayed, also had to check in bag because Carolyn bought Kayla and I stuff from bath
      &amp; body works</q>
   </div>`;

test('matches file names', () => {
   [
      `${name} - Text - 2012-01-29T04_57_07.html`,
      `${name} - Text - 2012-01-29T04_57_07.html`,
      `${name} - Voicemail - 2010-10-30T01_.html`,
      `${name} - Text - 2012-09-25T23_19_11.html`
   ].forEach(n => {
      expect(gv.filter(n)).toBe(true);
   });

   [
      `Family-Name.mbox`,
      `message.json`,
      `${name} - Voicemail - 2010-10-30T01_.mp3`
   ].forEach(n => {
      expect(gv.filter(n)).toBe(false);
   });
});

test('parses single message', () => {
   const html = parseHTML(raw);
   const msg = gv.parse(html);

   expect(msg).toEqual({
      source: Source.GoogleVoice,
      from: Relation.Other,
      on: new Date(2013, 5, 17, 18, 46, 0, 861),
      text: `Flight 2244, 15 minutes delayed, also had to check in bag because Carolyn bought Kayla and I stuff from bath
      & body works`
   });
});

test('parses file', async () => {
   const text = await readFileText(`${__dirname}/__mocks__/sms.html`);
   const messages = gv.process(text, 'mock');

   expect(messages).toBeInstanceOf(Array);
   expect(messages).toHaveLength(10);
   expect(messages[0].from).toBe(Relation.Other);
   expect(messages[1].from).toBe(Relation.Self);
   expect(messages[8].text).toBe('Superman? Home in 15');
   expect(messages[9].text).toBe('At park for a bit, signed papers');

   expect(messages).toMatchSnapshot();
});

test('parses file with single message', async () => {
   const text = await readFileText(`${__dirname}/__mocks__/sms2.html`);
   const messages = gv.process(text, 'shorty');

   expect(messages).toBeInstanceOf(Array);
   expect(messages).toHaveLength(1);
   expect(messages[0].from).toBe(Relation.Other);
   //expect(messages[0].on).toEqual(new Date(2012, 8, 25, 19, 19, 11, 272));
   expect(messages[0].text).toBe('At park for a bit, signed papers');

   expect(messages).toMatchSnapshot();
});
