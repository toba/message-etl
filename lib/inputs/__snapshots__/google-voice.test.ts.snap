// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parses file 1`] = `
Array [
  Object {
    "from": 2,
    "on": 2013-06-18T00:46:00.861Z,
    "source": 1,
    "text": "Flight 2244, 15 minutes delayed, also had to check in bag because Carolyn bought Kayla and I stuff from bath
        & body works",
  },
  Object {
    "from": 1,
    "on": 2013-06-18T00:48:20.247Z,
    "source": 1,
    "text": "Have good flying",
  },
  Object {
    "from": 2,
    "on": 2013-06-18T02:47:28.109Z,
    "source": 1,
    "text": "Landed",
  },
  Object {
    "from": 1,
    "on": 2013-06-18T02:47:58.180Z,
    "source": 1,
    "text": "We'll drive up the hill and await your text",
  },
  Object {
    "from": 2,
    "on": 2013-06-18T02:48:46.831Z,
    "source": 1,
    "text": "Ok",
  },
  Object {
    "from": 2,
    "on": 2013-06-18T03:00:17.481Z,
    "source": 1,
    "text": "Outside",
  },
  Object {
    "from": 1,
    "on": 2013-06-22T19:08:16.526Z,
    "source": 1,
    "text": "At Bonneville point",
  },
  Object {
    "from": 2,
    "on": 2013-06-22T19:09:13.122Z,
    "source": 1,
    "text": "So far so good, nice day for a ride",
  },
  Object {
    "from": 1,
    "on": 2013-06-26T23:51:05.573Z,
    "source": 1,
    "text": "Superman? Home in 15",
  },
  Object {
    "from": 2,
    "on": 2012-09-25T23:19:11.272Z,
    "source": 1,
    "text": "At park for a bit, signed papers",
  },
]
`;

exports[`parses file with single message 1`] = `
Array [
  Object {
    "from": 2,
    "on": 2012-09-25T23:19:11.272Z,
    "source": 1,
    "text": "At park for a bit, signed papers",
  },
]
`;
